_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2045 @@
+
+In summary, here are the key differences with live infrastructure configurations and the other two kinds of code we have
+discussed:
+
+- There is almost no form of automated testing you can implement in the predeployment stage. The only thing you can do
+  is perform a dry run and review the plan.
+
+- Similarly, there is no alternative environment where you can test the code manually during development; not even a
+  sandbox environment.
+
+- There are no release artifacts or tag with the live code. Everything is deployed immediately after reaching trunk.
+
+Now that we have an idea of what CI/CD pipelines look like, let's take a step back and define a threat model for CI/CD.
+This threat model will help us ensure that we implement the necessary security controls in these CI/CD pipelines so that
+we cover the common types of attack vectors for this type of workflow.
+
+
+[[threat_model_of_cicd]]
+=== Threat model of CI/CD
+
+To implement a CI/CD pipeline for infrastructure code, it is required that the ultimate entity or system running the
+infrastructure code has the permissions to deploy the infrastructure defined by code. Unfortunately, to support
+arbitrary CI/CD workflows, it is necessary to grant wide ranging permissions to the target environment. As such, it is
+important to consider ways to mitigate potential attacks against the various systems involved in the pipeline to avoid
+attackers gaining access to deploy targets, which could be catastrophic in the case of a breach of the production
+environment.
+
+Here we define our threat model to explicitly cover what attacks are taken into consideration in the design, as well as
+what attacks are __not__ considered. The goal of the threat model is to be realistic about the threats that are
+addressable with the tools available. By explicitly focusing attention on more likely and realistic threats, we can
+avoid overengineering and compromising the usability of the solution against threats that are unlikely to exist (e.g
+a 5 person startup with 100 end users is unlikely to be the subject of a targeted attack by a government agency).
+
+In this design, the following threat assumptions are made:
+
+- Attackers' goals are to gain access to an environment that they do not already have access to. Access to an
+  environment includes but is not limited to:
+    - The ability to read secrets that grant access to potentially sensitive data (e.g the database in prod
+      environment).
+    - Full access over all resources to cause damage to the business (e.g ability to delete the database and all its
+      backups in prod).
+
+- Attackers can originate from both external and internal sources (in relation to the organization).
+- External attacks are limited to those that can get full access to a CI environment, but not the underlying source
+  code. Note that __any__ CI/CD solution can likely be compromised if an attacker has access to your source code.
```

Comment:
Consider mentioning ways to protect your source code: e.g., protected branches, enforce MFA for your VCS, vulnerability scanning, etc.

Reply:
I wanted to leave mitigations for the production design, and wanted to focus on the threat model itself here. Otherwise, this section can get long with commentary.